### DISCUSSION TITLE
Wasm binary to wat conversion with type numbers but no renumbering?

### DISCUSSION THREAD

**Discussion by osa1** [WebAssembly/binaryen#5785] [Q&A] [answered]
I like wasm-opt's text output for debugging, but I can't find how to convert a .wasm to wat *with* type numbers but *without* renumbering.

This command:

```
wasm-opt -all --nominal test.wasm -o test.wat -S --name-types
```

generates types like

```
(type $type$14 (struct (field (mut i32))))
```

which is convenient because it makes the type hierarchy visible in the text and function types become simpler. However the number `14` here is not the actual struct index in the original program, binaryen renumbered the types.

To avoid renumbering I can use another pass that won't renumber the types:

```
wasm-opt -all --nominal test.wasm -o test.wat -S --vacuum
```

But that prints types like

```
(type ${mut:i32} (struct (field (mut i32))))
```

Now I can't see the struct type hierarchy in the output and I can't map a struct type index to the type in the output.

Is there a way to generate wat *with* type numbers but *without* renumbering?

**Comment by kripken** [comment, marked as answer]
I don't think there is a way to do that, sorry. `--name-types` picks arbitrary indexes for the types, and it's not possible to match them to the binary indexes, because Binaryen will reorder types in the binary writing. That is, in Binaryen IR we don't have type indexes, just types. When we write a binary we collect the types, sort by use frequency (to minimize binary size) and then use those indexes.
